sdks/java/core/src/main/java/org/apache/beam/sdk/testing/BigqueryMatcher.java

```diff
@@ -123,22 +123,42 @@ Credential getDefaultCredential(HttpTransport transport, JsonFactory jsonFactory
     }
 
     if (credential.createScopedRequired()) {
-      Collection<String> bigqueryScope = BigqueryScopes.all();
+      Collection<String> bigqueryScope =
+          Lists.newArrayList(BigqueryScopes.CLOUD_PLATFORM_READ_ONLY);
       credential = credential.createScoped(bigqueryScope);
     }
     return credential;
   }
 
-  String hashing(List<TableRow> rows) {
-    if (rows == null || rows.isEmpty()) {
-      LOG.warn("Invalid data for hashing. Empty checksum is returned.");
-      return "";
-    }
-    List<HashCode> hashCodes = new ArrayList<>();
+  private String generateHash(List<TableRow> rows) {
+    List<HashCode> rowHashes = Lists.newArrayList();
     for (TableRow row : rows) {
-      hashCodes.add(Hashing.sha1().hashString(row.toString(), StandardCharsets.UTF_8));
+      List<HashCode> cellHashes = Lists.newArrayList();
+      for (TableCell cell : row.getF()) {
+        cellHashes.add(Hashing.sha1().hashString(cell.toString(), StandardCharsets.UTF_8));
+      }
+      rowHashes.add(Hashing.combineUnordered(cellHashes));
     }
-    return Hashing.combineUnordered(hashCodes).toString();
+    return Hashing.combineUnordered(rowHashes).toString();
+  }
+
+  private QueryResponse queryWithRetries(Bigquery bigqueryClient) {
```

Comment:
This is the right idea, as it turns out there's a util class which gives you a lot of this logic for free, as well as flexibility during testing. Take a look at Sleeper and BackOff here:

https://github.com/apache/incubator-beam/blob/900980246b3ee3bdda8009672d5092a9260f06ba/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java#L192